ui: Add guest user suffix

AnkurPrabhu opened the pull request:
This commit is for adding guest user suffix in the UI if the user is a guest user and if realm_enable_guest_user_indicator setting is True. 
<!-- See README for documentation, or ask in #zulip-terminal if unclear -->
### What does this PR do, and why?
This PR Implements #1444  
The approach here is to update the function user_name_from_id to return the user's name with the suffix (guest) if the condition satisfies and calling the function in places where we display user's name.
The approach for Right column view is different from others we just pass the required arguments (realm_enable_guest_user_indicator  and is_guest) to UserButton and add suffix to the users name depending upon the arguments



### Outstanding aspect(s)    <!-- DELETE SECTION IF EMPTY -->
<!-- In what ways is this not fully implemented/functioning? Compared to a discussion/issue? -->
<!-- Do you not understand something? Are you unsure about a certain approach? Want feedback? -->
- [ ] wanted to know if my approach is right 
- [ ] other scenarios where the display change needs to be done
- [ ] will fix the test cases once approach is approved 

### External discussion & connections
<!-- [x] all that apply, specifying topic and adding numbers after # for issues/PRs -->
- [x] Discussed in **#zulip-terminal** in `Add guest user suffix #T1444 #T1460`
- [ ] Fully fixes #
- [x] Partially fixes issue #1444
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
<!-- [x] all that apply -->
- [x] Manually - Behavioral changes
- [x] Manually - Visual changes
- [ ] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [ ] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [ ] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the  motivation and reasoning for the change
- [ ] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [ ] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes    <!-- DELETE SECTION IF NO VISUAL CHANGE -->
<!-- Zulip tips at https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html -->
<!-- For video, try asciinema; after uploading, embed using
[![yourtitle](https://asciinema.org/a/<id>.png)](https://asciinema.org/a/<id>)
-->
<!-- NOTE: Attached videos/images will be clearer from smaller terminal windows -->
<img width="161" alt="Screenshot 2024-01-13 at 2 15 00 AM" src="https://github.com/zulip/zulip-terminal/assets/85862184/8c68b4ef-ba60-4bdc-933f-4d85686aaaae">
<img width="499" alt="Screenshot 2024-01-13 at 2 15 07 AM" src="https://github.com/zulip/zulip-terminal/assets/85862184/fbcb88e5-f80b-4c8b-adc1-81e7d53c602f">
<img width="272" alt="Screenshot 2024-01-13 at 2 15 24 AM" src="https://github.com/zulip/zulip-terminal/assets/85862184/6257e938-b1ed-406d-bb13-5ed907a4ffc3">
<img width="282" alt="Screenshot 2024-01-13 at 2 15 39 AM" src="https://github.com/zulip/zulip-terminal/assets/85862184/418c1bf2-fe76-4c7b-bf1d-f263b4d9002c">
<img width="470" alt="Screenshot 2024-01-13 at 2 20 13 AM" src="https://github.com/zulip/zulip-terminal/assets/85862184/9521bded-2de6-4137-8f51-ef02ab646f87">

neiljp reviewed (commented):
@AnkurPrabhu Thanks for working on this :) :+1:

The UI you have in your initial screenshots look like a good starting point 👍 The main concern would be to have consistent styling for the guest suffix, and certainly a space from the name. Specifically, we need it to be distinct from the user name, so it is clear that it is not part of the user name! A new style may be warranted for this purpose, since keeping it distinct will change with theme (see the themes main file and individual themes).

I've noted inline some general points and also regarding keeping the model and UI logic separate from each other.

While it seems like this is a relatively small change, particularly when including tests you will likely find it clearer to split this into smaller commits and address the changes to the model and each UI area separately. That makes it easier to merge good parts earlier, if the commits are relatively independent and can be reordered, or set up with simpler commits first. The user list changes look like they may be more complex, for example.

We may need to decide in a followup or in this PR what to do if text wraps. I think the server had a similar issue with the user list at one point - eg. do we prioritize the user name or the guest marker, or both? This also applies if the active user is a guest!

AnkurPrabhu commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -287,6 +289,12 @@ def __init__(
             count=count,
         )
+
+        if realm_enable_guest_user_indicator and is_guest:
+            self.suffix_style = "current_user"
+            self.suffix_text = "(guest)"
+            self.update_widget()
```
the web app does not contain '(you)' for the current user but on hover it overrides the guest suffix take a look at the screenshot attached, Ill put this down the code so that '(you)' gets overridden if you want me to add guest for the active user as well.  
<img width="469" alt="Screenshot 2024-01-16 at 7 42 13 PM" src="https://github.com/zulip/zulip-terminal/assets/85862184/0513a565-4284-47c6-94bd-e79fc17fd2e5">

neiljp replied:
Yeah, this will be part of the challenge with how to render this in the user list. I'm not sure there's an obvious solution given our platform constraints, but isolating these changes to their own commit will make this easier to examine separately.

AnkurPrabhu commented on zulipterminal/ui_tools/views.py:
```diff
@@ -729,7 +730,12 @@ def users_view(self, users: Any = None) -> Any:
             unread_count = self.view.model.unread_counts["unread_pms"].get(
                 user["user_id"], 0
             )
+            realm_enable_guest_user_indicator = self.model.initial_data[
+                "realm_enable_guest_user_indicator"
+            ]
```
Hey for this ill use 
realm_enable_guest_user_indicator = self.model.initial_data.get(
                "realm_enable_guest_user_indicator", None
            )
so that it returns None when not present, also are you suggesting not to use this here ?

neiljp replied:
Right, that will avoid the exception.

As per another comment, in (new) code we are aiming to have server changes (different versions) isolated in the model, so that UI code isn't dependent on that.

AnkurPrabhu commented on zulipterminal/ui_tools/messages.py:
```diff
@@ -713,7 +718,8 @@ def main_view(self) -> List[Any]:
                     ("pack", urwid.Text(text["status"])),
-                    ("weight", 10, urwid.Text(text["author"])),
+                    ("pack", urwid.Text(text["author"])),
+                    ("weight", 10, urwid.Text(text["suffix"])),
```
added like a new widget

AnkurPrabhu commented:
Hey @neiljp i have tried a new approach in this one pls let me know your thoughts also as you told to split it in different prs this pr will be related to adding the guest suffix in the messages box (changes related to header and recipients).
For displaying the guest suffix in the right column that will be a separate pr but the changes in the button have been added here as I wanted to know your thoughts.

neiljp commented on zulipterminal/model.py:
```diff
@@ -1262,12 +1265,17 @@ def user_name_from_id(self, user_id: int) -> str:
+        if guest_user_indicator and is_guest:
+            return name + "(guest)"
+        else:
+            return name
```
I mainly meant the name and the suffix as the two parts.

In terms of 'where the suffix goes', the styling and placement of the suffix is not something that the model should be handling - that is an issue for the UI. The model should only be concerned with providing an answer to "should I show a guest suffix for this user".

neiljp commented on zulipterminal/themes/zt_dark.py:
```diff
@@ -68,6 +68,7 @@
     'task:error'      : (Color.WHITE,               Color.DARK_RED),
     'task:warning'    : (Color.WHITE,               Color.BROWN),
     'ui_code'         : (Color.BLACK,               Color.WHITE),
+    'guest_suffix'    : (Color.WHITE,               Color.BLACK),
```
This is one place where a style can be added to a theme, but note there are others - it will need adding in each place.

For example, look for one of the other style names throughout the code.

See previous style-adding commits in the git log for examples - we tend to add a style in all places in a single commit.

AnkurPrabhu replied:
yes i saw there are a lot of styles i just wanted to show this to know if im on the right track or not, will add it in all the files

AnkurPrabhu replied:
I have done it in all the files right ?

neiljp commented on zulipterminal/ui_tools/utils.py:
```diff
@@ -30,6 +31,10 @@ def create_msg_box_list(
     last_msg = last_message
     muted_msgs = 0  # No of messages that are muted.
     for msg in message_list:
+        if model.initial_data.get("realm_enable_guest_user_indicator", None):
+            user_info = model.get_user_info(msg["sender_id"])
+            role = user_info["role"]
+            msg["sender_role"] = ROLE_BY_ID[role]["name"]
```
I understand what you're trying to achieve here, ie. caching the name of the role in messages.

This does work, but the messages data structure already potentially has fields dedicated to various roles, as well as the main role field in late versions.

AnkurPrabhu replied:
so ill make a separate method which will make the two parts of the name the name and suffix 
by default the suffix will be "" (blank)  so that way we do not have to check any setting from model and wherever we are displaying names we will call this method ? does this work ?

neiljp replied:
Yes, that may be useful, though it may depend on how the UI works in different places, eg. our custom buttons do not accept the same input as a simple text box. If you apply it to one area first, then you can just have the logic at that position.

The logic may be quite small, once you have the new method in the model, to indicate if a guest suffix should be shown for a given user id.

neiljp commented:
@AnkurPrabhu Thanks for following up. Unfortunate I didn't fetch the previous version, so it's difficult for me to identify precisely what you've changed in this version. I've left some notes inline, as well as responses to earlier comments.

Please note:
- I was mainly referring to multiple commits; these can be with the same PR if they work together
- some suffix work can certainly be left to another PR, but the foundational work likely wants to be in the first PR
- I know you left the button changes in, but you could still keep these in the PR, but in a different commit :)

For the next step, I'd suggest splitting your work here into multiple commits in this PR, and thinking how you might write tests for each new code. Firstly I'd advise adding a model method like I mentioned in a response to an earlier comment, and then adapting the rest of your code to use that method.

neiljp reviewed (commented):
@AnkurPrabhu I left my main comments in another comment, which refers to these comments too :)

AnkurPrabhu commented on zulipterminal/model.py:
```diff
@@ -143,6 +143,9 @@ def __init__(self, controller: Any) -> None:
             "zulip_version",
+            # realm_enable_guest_user_indicator is present  in
+            # POST /register from Feature level 216.
+            "realm_enable_guest_user_indicator",
```
yes its not necessary ill remove it

AnkurPrabhu commented:
@neiljp this i have updated the pr with the latest changes:
1) have added style to all the current themes
2) currently showing guest user suffix at
- message header
- message recipients
- right column (button)
- message info view 
- message stream members view
- user info view 
let me know if I am missing the suffix at any place 

all these felt like similar changes only so once you approve my current approach will start writing tests for the same .
also i was thinking of making a constant in the model like guest_suffix= "( guest)" so that we can use this everywhere instead if writing it as a string everywhere... @neiljp  thoughts ?

AnkurPrabhu commented:
@neiljp i have fixed all the existing tests can you tell me how to fix this gitlint one ?

neiljp commented:
@AnkurPrabhu The test error appears to be due to your branch not including the setup code for the gitlint in CI. This should resolve if you fetch the latest `main` branch and rebase this branch against the updated `main` which has the appropriate commits to install gitlint in CI.

However, I suspect you will have gitlint errors locally too. See the docs, but eg try something like `gitlint --commits HEAD~3..` (or with `main..`, once you're up to date with main).

AnkurPrabhu commented:
> @neiljp this i have updated the pr with the latest changes:
> 
> 1. have added style to all the current themes
> 2. currently showing guest user suffix at
> 
> * message header
> * message recipients
> * right column (button)
> * message info view
> * message stream members view
> * user info view
>   let me know if I am missing the suffix at any place
> 
> all these felt like similar changes only so once you approve my current approach will start writing tests for the same . also i was thinking of making a constant in the model like guest_suffix= "( guest)" so that we can use this everywhere instead if writing it as a string everywhere... @neiljp thoughts ?

@neiljp you missed this i guess if the changes are good ill write tests and we can close this

neiljp commented:
@AnkurPrabhu Those locations look good; if we miss one then it can be added later.

I agree that extracting a common constant would be good, but note again that this is not something that would belong in the model, for the same reasons as before.

neiljp commented on zulipterminal/model.py:
```diff
@@ -2071,3 +2071,17 @@ def poll_for_events(self) -> None:
                         )
+
+    def show_guest_suffix(self, user_id: int) -> bool:
+        """
+        Returns True if the realm_enable_guest_user_indicator
+        setting is enabled and the user is a guest
+        """
```
A docstring is for users of the function, not a comment as to how it works.

If you feel there is a need for a description, beyond the name of the function, then please adjust the docstring - but it should not describe the implementation.

neiljp commented on zulipterminal/model.py:
```diff
@@ -2071,3 +2071,17 @@ def poll_for_events(self) -> None:
+        setting is enabled and the user is a guest
+        """
+        user = self.get_user_info(user_id)
+        assert user is not None
```
Based on existing model code, let's raise a RuntimeError here.

We should move towards better exceptions for these cases in future though.

neiljp commented on zulipterminal/model.py:
```diff
@@ -2071,3 +2071,17 @@ def poll_for_events(self) -> None:
                         self.controller.raise_exception_in_main_thread(
                             sys.exc_info(), critical=False
                         )
+
```
You've split commits more here, but note the commit text - you're updating model.py, so 'ui' doesn't apply to the 'area' for the title.

neiljp reviewed (commented):
@AnkurPrabhu I just left some feedback in addition to answering your questions.

The new model method is in the right direction, and the split commits are an improvement :+1:

However, please do split the UI changes into separate ones for each UI part that you update.

AnkurPrabhu commented on zulipterminal/model.py:
```diff
@@ -2071,3 +2071,17 @@ def poll_for_events(self) -> None:
+        assert user is not None
+        role = user["role"]
+
+        return (
+            self.initial_data.get("realm_enable_guest_user_indicator", None)
+            and ROLE_BY_ID[role]["name"] == ROLE_BY_ID[600]["name"]
+        )
```
I am using ( and ) condition here so if the first one returns false it will never calculate the other half of the condition right ?  
yes it looks complicated ill make it more readable

neiljp replied:
The expression won't calculate the later part if the former is false, but `role` itself will still be calculated in the part above this, independently of that.

AnkurPrabhu commented:
@neiljp I have split it based on ui. 
I have not split the message info view as i thought they are part of the message itself.
I Have fixed all the existing tests. 

@neiljp  can you guide me on what are the next steps or changes i should work on

mounilKshah commented on zulipterminal/ui_tools/messages.py:
```diff
@@ -91,16 +92,24 @@ def __init__(self, message: Message, model: "Model", last_message: Any) -> None:
-                        if recipient["email"] != self.model.user_email
-                    ]
-                )
+                self.recipients_names = ""
+                recipients = []
+                for recipient in self.message["display_recipient"]:
+                    if recipient["email"] != self.model.user_email:
+                        recipient_name = self.recipients_names = recipient["full_name"]
+                        if self.model.show_guest_suffix(user_id=recipient["id"]):
+                            recipient_name += " (guest)"
+                        recipients.append(recipient_name)
+                self.recipients_names = ", ".join(recipients)
```
This works fine but you can also maintain the original python comprehension and only add the suffix to the `recipient_name` with a conditional, similar to how it's done above in the private message conditional block

mounilKshah reviewed (commented):
Thank you for updating the PR @AnkurPrabhu 
This works fine on my system. I think the commit titles and messages for all commits can be modified in accordance with the Zulip commit standards.

I had a few suggestions and have mentioned them in the review.

AnkurPrabhu committed:
model: Add guest user suffix.

This commit is for adding a function is_guest_user() in the model
which will validate if a user is a guest user and
realm_enable_guest_user_indicator setting is True.

AnkurPrabhu committed:
themes: Add guest_suffix styling to themes.

This commit is for adding the style for guest suffix for all the themes
currently present

AnkurPrabhu committed:
ui: Add guest user suffix.

This commit is for adding guest user suffix in the right column of the
UI. The approach here is to update the username by passing
the value returned by model.is_guest_user() to the UserButton
and depending upon the argument passed we show the suffix.

AnkurPrabhu committed:
ui: Add guest user suffix.

This commit is for adding guest user suffix in the UI if the user is
a guest user and if realm_enable_guest_user_indicator setting is True.
The approach here is to update the username by adding the guest
suffix  if model.is_guest_user() returns True.

neiljp reviewed (commented):
@AnkurPrabhu Thanks for making the changes :+1:
(as noted in the stream, I started this yesterday, but had issues on my reviewing computer)

I've yet to test this on a live server with guest users - have you done so?

I've left some followup comments where you or I missed something.

In addition to the inline comments:
- Somewhere we discussed adding the `(guest)` string as a constant? (https://github.com/zulip/zulip-terminal/pull/1460#issuecomment-1925002202) The `ui_mappings.py` file seems like a reasonable place for this; it can be in it's own commit, and then we can squash it with the first UI commit that gets merged.
- It's good to see the user list changes in a separate commit, but could you separate the changes to each other UI part separately too? I expect they should be independent of each other, and that would make it much easier to see the changes associated with each, test them separately, reorder them, or work on them later - and the commit title/body can vary for each to explain the feature you are adding in that specific commit.
- Each commit title should be able to be relatively standalone, and describe what it does itself, not the PR overall; this might be just due to the current commits not being updated, but that also helps with review - it explains each commit as part of the set.

The user column looks like it is going to be the more complex part, so that commit might be better at the end. This may need to be a symbol, as for text like elsewhere it needs more space, and so might be dependent upon a dynamic larger column width if the window is wide enough, which someone looked into working on a while back.
